### ISSUE TITLE
ReentrentGrainWithState test is failing

### ISSUE THREAD

**Issue by sergeybykov** [dotnet/orleans#1183]
`ReentrentGrainWithState` has been recently migrated to GitHub and it seems to be failing consistently with the `MemoryStorage` provider throwing an etag violation exception.

The logic in the test isn't self-evident, so it's unclear if this is a test or product issue.

**Comment by gabikliot** [comment]
Will take a look. I just recently changed the etag handling in memory storage.

**Comment by gabikliot** [comment]
Yes, of course, this Reentrant grain was doing concurrent writes, which would cause etag violations, as expected. The fix, in this test and in general in any Reentrant grain with state is to **serialize state writes: never issue more than one write concurrently**. Very easy to implement.

Fixed via #1185.

**Comment by gabikliot** [comment]
BTW, I want to add something about this test, which I think people might find useful.
This test is sort of ambiguous: on the one hand, it's a mess! If you look at it, it is very hard to follow and understand what is going on, it tests multiple concerns together, its hard to maintain and debug. 

Yet, it is a VERY good test. It really stresses the concurrency of the Reentrant grain. There are tons of interleaving, and a lot of concurrent state writes. Its test coverage of the hard-to-hit concurrent interleavings is very good! If you look at my fix #1185, I did not put the `while` condition there initially. Of course, I put `if` first. And the test still failed, on iteration 86! And that was a really hard to hit interleaving.

So I think of course we also need (and we have) more isolated, single test-point tests, but we also need to keep and write more of those complicated, "not unit tests but real integration" tests.
